### PR TITLE
Puma

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Switches webserver from Unicorn to Puma.

#### What should we test?

Before deploying to staging, the server will need to be provisioned by a dev using [this branch](https://github.com/openfoodfoundation/ofn-install/pull/752) of ofn-install.

The app should basically work as before, but be a bit snappier. 

One thing that's worth testing is timeouts on really long requests (like reports with tons of data). I'm 99% sure it'll work exactly as before with the settings I've added, but timeout handling is slightly different in Puma so it's worth a check to make sure.

The test suite will also be using Puma as well to run all controller and feature specs, more details in [this comment](https://github.com/openfoodfoundation/openfoodnetwork/pull/7746#issuecomment-888250983).

#### Release notes

Switched app server from Unicorn to Puma

Changelog Category: Technical changes

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
I did a bit of load testing here on Katuma Staging with **Siege** before and after (Unicorn vs Puma). Interesting results! 

### Memory Use

First off, with Unicorn (2 workers) the total memory usage on Katuma Staging was around ~1.2 GB. With Puma (also 2 workers), the total memory use (for the whole server) dropped to ~850MB.

### Concurrency

Whereas **Unicorn** has a single thread per worker, multi-threaded **Puma** (in this modest configuration) has 3 threads per worker, so it can theoretically handle triple the amount of concurrent requests with the same number of workers (and whilst using less memory!).

Here's the results from Siege. This test emulates 25 user concurrently hitting the site (for 30 seconds) and collates the results.

**Unicorn:**
```
** SIEGE 4.0.4
** Preparing 25 concurrent users for battle.
The server is now under siege...
Lifting the server siege...
Transactions:		          62 hits
Availability:		      100.00 %
Elapsed time:		       29.45 secs
Data transferred:	        7.22 MB
Response time:		        4.99 secs
Transaction rate:	        2.11 trans/sec
Throughput:		        0.25 MB/sec
Concurrency:		       10.50
Successful transactions:          62
Failed transactions:	           0
Longest transaction:	       26.41
Shortest transaction:	        0.49
```
**Puma:**
```
** SIEGE 4.0.4
** Preparing 25 concurrent users for battle.
The server is now under siege...
Lifting the server siege...
Transactions:		         121 hits
Availability:		      100.00 %
Elapsed time:		       29.35 secs
Data transferred:	       19.74 MB
Response time:		        5.71 secs
Transaction rate:	        4.12 trans/sec
Throughput:		        0.67 MB/sec
Concurrency:		       23.55
Successful transactions:         121
Failed transactions:	           0
Longest transaction:	       24.97
Shortest transaction:	        0.23
```

So the total number of requests served (with 25 concurrent users in 30 seconds) changed from **62** to **121**. Requests per second went from **2.11** to **4.11**. "Longest transaction" and "Shortest transaction" are also significantly lower with Puma. This is with a ~25% smaller memory footprint :point_up: !

**Comment by Matt-Yorkley** [comment]
On bigger instances that currently have 3-5 Unicorn workers, I think we could potentially cut memory use by 35-50% whilst increasing throughput substantially. We'd then have a lot of room to look at increasing either worker count or thread count per worker if extra capacity was needed in the future...

**Comment by lin-d-hop** [comment]
@Matt-Yorkley Are there ANY downsides here?
Shall we just go for it?

**Comment by Matt-Yorkley** [comment]
It's basically a no-brainer, all modern Rails apps use Puma by default. I think maybe we can wait until this week's release is drafted tomorrow, then we can test/merge and we'll have a week of running it in Staging servers that will give us a good amount of insight as to whether or not there are any issues.

**Comment by lin-d-hop** [comment]
Sounds good. I notice in the PR that there isn't really any Puma configuration added.
Is that because Puma is so excellent out of the box?
Or is there a chance we'll need to be on standby to firefight config issues after deployment?

**Comment by Matt-Yorkley** [comment]
The config is in ofn-install here: https://github.com/openfoodfoundation/ofn-install/pull/752

**Comment by Matt-Yorkley** [comment]
One of the things this PR does is switch our test suite over to using Puma instead of the default Webrick (which is old and clunky and has a few problems sometimes). So that means a large number of our specs are already being run through Puma in this PR (which has a green build :heavy_check_mark:).

**Comment by Matt-Yorkley** [comment]
On the rollout side of things the important question is: will the switchover require downtime?

And the answer is: yes... unless you do some DevOps magic using nginx connection fallbacks and unix sockets... which looks a bit like this: https://github.com/openfoodfoundation/ofn-install/pull/752/commits/8098743ba50e3957ab317a000fed7f374fd9c2d3

Then no :wink:

**Review by andrewpbrett — APPROVED** [review]
💯

File: config.ru [hunk 1]
```diff
@@ -2,18 +2,5 @@
...
-  # - grown in memory usage to between 700 and 850 MB
-  require 'unicorn/worker_killer'
-  use Unicorn::WorkerKiller::MaxRequests,
-      ENV.fetch('UWK_REQS_MIN', 25_000).to_i,
-      ENV.fetch('UWK_REQS_MAX', 30_000).to_i
-  use Unicorn::WorkerKiller::Oom,
-      ( ENV.fetch('UWK_MEM_MIN', 700).to_i * (1024**2) ),
-      ( ENV.fetch('UWK_MEM_MAX', 850).to_i * (1024**2) )
-end
-
```
Comment (jibees) [inline config.ru:17]: 👍

**Review by jibees — APPROVED** [review]
👌

**Comment by jibees** [comment]
Should we move this one to `Dev Test Ready`?

**Comment by filipefurtad0** [comment]
The deployment failed after provisioning, with this reference to Unicorn:
```

RUNNING HANDLER [restart unicorn] **********************************************
fatal: [staging.openfoodnetwork.org.au]: FAILED! => {"changed": false, "msg": "Unable to restart service unicorn_openfoodnetwork: Job for unicorn_openfoodnetwork.service failed because the control process exited with error code. See \"systemctl status unicorn_openfoodnetwork.service\" and \"journalctl -xe\" for details.\n"}
Friday 30 July 2021  17:19:47 +0000 (0:00:04.010)       0:12:02.334 *********** 
```

**Comment by Matt-Yorkley** [comment]
I forgot to mention it needed manual deployment with openfoodfoundation/ofn-install#752

It's deployed now on Aus Staging :ok_hand:

**Comment by Matt-Yorkley** [comment]
Just did some quick dev-tests in relation to timeouts as they work a bit differently with puma. Although nginx halts the request from the user's perspective, puma carries on processing the request after it's passed the timeout limit, which isn't ideal. I've added some explicit timeout handling via Rack so long requests now stop correctly and free up the thread/worker as expected :ok_hand:

**Comment by RachL** [comment]
@Matt-Yorkley I've ran a basic release test I couldn't find anything unusual. I guess this is ready to go!

**Pull request merged**